Don't show the tap icon for requests from sources that are not meshed

alpeb opened the pull request:
Fixes #1723

For testing, I injected all the emojivoto deployments but vote-bot. You can see the requests coming from vote-bot are not tappable anymore:

![image](https://user-images.githubusercontent.com/554287/51622118-f0649f80-1f03-11e9-804d-27a5e39cc63b.png)

grampelberg commented:
Why can't we tap the vote-bot requests?

alpeb commented:
@grampelberg From what I understand, the `tap` API accepts a resource and optionally a target (TO) resource. Currently it doesn't have the option to specify a FROM.
So in order to tap requests from `vote-bot` to `web`, we're not tapping `web` but actually `vote-bot`:

`linkerd tap deployment/vote-bot --namespace emojivoto --to deployment/web --to-namespace emojivoto`

That requires `vote-bot` to be meshed, which is not in the example above.

grampelberg commented:
`--from` sounds like a pretty valuable filter for tap. @adleong you wouldn't happen to have any history there would you?

adleong commented:
I believe the problem is that we don't have source metadata for requests.  We can tap requests coming into the web deployment, but we can't tell which of them came from the vote-bot.

rmars commented on web/app/js/components/TopModule.jsx:
```diff
@@ -237,7 +239,8 @@ class TopModule extends React.Component {
     if (d.base.proxyDirection === "INBOUND") {
-      this.props.updateNeighbors(d.requestInit.source, _get(d, "requestInit.sourceMeta.labels"), null);
+      let unmeshedSources = this.props.updateNeighbors(d.requestInit.source, _get(d, "requestInit.sourceMeta.labels"), null);
+      topResults[eventKey].meshed = !_has(unmeshedSources, d.base.source.owner);
```
this check doesn't quite work when the resource we're doing the Top on is a `pod`. E.g. with web meshed, if you go to the Web's Resource Detail page for pod (e.g. http://localhost:8084/namespaces/emojivoto/pods/web-6b4675554d-ldtzm) you'll see that the tap icon continues to appear for all rows.

I think you'd need something like:


```suggestion
      let isPod = this.props.query.resource.split("/")[0] === "pod";
      if (isPod) {
        topResults[eventKey].meshed = !_has(unmeshedSources, "pod/" + d.base.source.pod);
      } else {
        topResults[eventKey].meshed = !_has(unmeshedSources, d.base.source.owner);
      }
```

alpeb replied:
Ok, understood :+1:

alpeb commented:
@rmars thanks for the detailed feedback. You provided above a diff for `Top.jsx`, but your comment states we should move `updateNeighborsFromTapData` into `TopModule`. Was that a mistake, or am I totally misunderstanding?
On the other hand, I think we still need `updateNeighborsFromTapData` to be in `ResourceDetail.jsx` because `<Octupus>` uses `unmeshedSources` which is populated as a side-effect of having that function live in that class...

rmars commented:
@alpeb yes! indeed! I was saying that once we implemented the above functionality in Top (i.e. my diff), we would notice that both `Top` and `ResourceDetail` pages would have similar `updateNeighborsFromTapData` methods, and that maybe we could then consider moving the `updateNeighborsFromTapData` functionality into `TopModule` so that the logic is in one place. 

(`TopModule` would still need to pass the unmeshedSources it collected to its parent component, though, since ResourceDetail needs it to render the Octopus graph. But the calling of the `processNeighborData` could be put in one place.) It might be that it doesn't make sense to consolidate them, though, and leaving two `updateNeighborsFromTapData` would be fine too!

alpeb commented:
Thanks for the clarification @rmars.

I moved `updateNeighborsFromTapData` into `TopModule` and now the Top and ResourceDetail pages are properly hiding the tap icons.

Now `<Top>` is receiving as an attribute the `updateUnmeshedSources` function instead of `updateNeighbors`, which is used to propagate the `unmeshedSources` object into the parent components.                                                                                             

The logic for updating the `meshed` property has been updated to take into account pods, like you suggested.

Let me know what you think :thinking:

alpeb commented on web/app/js/components/TopModule.jsx:
```diff
@@ -97,7 +98,6 @@ class TopModule extends React.Component {
 
   onWebsocketRecv = e => {
     this.indexTapResult(e.data);
-    this.props.updateNeighbors(e.data);
```
It appears this wasn't doing anything (since not sending anything as a secong arg) so I removed it.

rmars replied:
good catch!

rmars commented on web/app/js/components/TopModule.jsx:
```diff
@@ -26,8 +27,8 @@ class TopModule extends React.Component {
-    updateNeighbors: PropTypes.func,
-    updateTapClosingState: PropTypes.func
+    updateTapClosingState: PropTypes.func,
+    updateUnmeshedSources: PropTypes.func
```
since we now depend on this to show the tap icon in the TopTable, I think we should make this `isRequired`

alpeb replied:
When building  `<TopModule>` in the `/top` page I'm not passing any function because there's no octopus that needs an `unmeshedSources` object to be available.

rmars replied:
ah, right! I thought we needed this to keep track of the unmeshed upstreams for the table, but we're keeping track of them in TopModule

rmars commented on web/app/js/components/TopModule.jsx:
```diff
@@ -253,6 +261,14 @@ class TopModule extends React.Component {
+    // is very costly and causes the page to freeze up
+    let resourceType = _isNil(this.props.query.resource) ? "" : this.props.query.resource.split("/")[0];
+    this.unmeshedSources = processNeighborData(source, sourceLabels, this.unmeshedSources, resourceType);
+    if (this.props.updateUnmeshedSources) {this.props.updateUnmeshedSources(this.unmeshedSources);}
```
you don't need this check here, since the current default value of this func is `_noop` so it won't do anything. but also, if we make this prop required, we won't need to worry about checking it here!

alpeb replied:
ah that makes sense :+1:

rmars commented on web/app/js/components/util/TapUtils.jsx:
```diff
@@ -354,7 +354,9 @@ export const tapLink = (d, resourceType, PrefixedLink) => {
 
-  if (_has(d.sourceLabels, resourceType)) {
+  if (!d.meshed) {
+    return null;
```
TIOLI we could return a greyed out, non-clickable Tap icon for resources that aren't tappable

alpeb replied:
will do :sparkles:

alpeb commented:
@rmars please check my comment about the required function. I've pushed your other suggestions, including the grayed-out icon, which looks much better. Not sure though what are the project's standards regarding css...

rmars commented on web/app/js/components/util/TapUtils.jsx:
```diff
@@ -352,14 +352,19 @@ export const srcDstColumn = (d, resourceType, ResourceLink) => {
 export const tapLink = (d, resourceType, PrefixedLink) => {
   let namespace = d.sourceLabels.namespace;
-  let resource = "";
+  let resource;
```
could we set namespace/resource to "" here instead, so we don't have to deal with the subsequent cases?

```suggestion
  let namespace = d.sourceLabels.namespace || "";
  let resource = "";
```

That also kind of got me thinking, that instead of forcing these to be "" to get around the fact that namespace/resource are required by TapLink, that we could instead make those prop types be not required, but have a default value of "".

Another random thought I had was maybe we should instead of doing that, add a `disabled` prop that would control whether the tap icon was greyed out. Just trying to think of whatever is clearest, it's not that important though :)

alpeb replied:
I like the idea of `disable`, the intention would be clearer :+1:

alpeb commented:
@rmars In my last push I addressed your comment, and I also fixed a missing scenario where the tap icon should be grayed-out (when `d.base.source.owner` is empty) :microscope:

rmars commented on web/app/js/components/TapLink.jsx:
```diff
@@ -24,6 +24,7 @@ const TapLink = ({PrefixedLink, namespace, resource, toNamespace, toResource, pa
 };
 
 TapLink.propTypes = {
+  disabled: PropTypes.bool.isRequired,
```
TIOLI: I think `disabled` shouldn't be required; and that it should default to `false` since that'd probably be the most common use case; this doesn't matter much for this instance since `tapLink` is the only place we actually create a `TapLink`....

```
TapLink.defaultProps = {
  disabled: false
}
```

alpeb replied:
no problemo :+1:

rmars commented on web/app/js/components/TopModule.jsx:
```diff
@@ -244,7 +245,7 @@ class TopModule extends React.Component {
         topResults[eventKey].meshed = !_has(this.unmeshedSources, "pod/" + d.base.source.pod);
       } else {
-        topResults[eventKey].meshed = !_has(this.unmeshedSources, d.base.source.owner);
+        topResults[eventKey].meshed = !_isEmpty(d.base.source.owner) && !_has(this.unmeshedSources, d.base.source.owner);
```
good catch!
